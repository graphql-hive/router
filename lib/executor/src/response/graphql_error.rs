use serde::{de, Deserialize, Deserializer, Serialize};
use sonic_rs::Value;
use std::fmt;

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GraphQLError {
    pub message: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub locations: Option<Vec<GraphQLErrorLocation>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub path: Option<Vec<GraphQLErrorPathSegment>>,
    pub extensions: Option<Value>,
}

impl From<String> for GraphQLError {
    fn from(message: String) -> Self {
        GraphQLError {
            message,
            locations: None,
            path: None,
            extensions: None,
        }
    }
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct GraphQLErrorLocation {
    pub line: usize,
    pub column: usize,
}

#[derive(Clone, Debug, Serialize)]
pub enum GraphQLErrorPathSegment {
    String(String),
    Index(usize),
}

impl<'de> Deserialize<'de> for GraphQLErrorPathSegment {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        struct PathSegmentVisitor;

        impl<'de> de::Visitor<'de> for PathSegmentVisitor {
            type Value = GraphQLErrorPathSegment;

            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
                formatter.write_str("a string or an integer for a GraphQL path segment")
            }

            fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
            where
                E: de::Error,
            {
                Ok(GraphQLErrorPathSegment::String(value.to_owned()))
            }

            fn visit_string<E>(self, value: String) -> Result<Self::Value, E>
            where
                E: de::Error,
            {
                Ok(GraphQLErrorPathSegment::String(value))
            }

            fn visit_u64<E>(self, value: u64) -> Result<Self::Value, E>
            where
                E: de::Error,
            {
                Ok(GraphQLErrorPathSegment::Index(value as usize))
            }

            fn visit_i64<E>(self, value: i64) -> Result<Self::Value, E>
            where
                E: de::Error,
            {
                if value < 0 {
                    return Err(E::custom(format!(
                        "path segment must be a non-negative integer, but got {}",
                        value
                    )));
                }
                Ok(GraphQLErrorPathSegment::Index(value as usize))
            }
        }

        deserializer.deserialize_any(PathSegmentVisitor)
    }
}
