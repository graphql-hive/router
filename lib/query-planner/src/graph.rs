use std::{
    collections::HashMap,
    fmt::{Debug, Display},
};

use graphql_parser_hive_fork::schema::ParseError;
use graphql_tools::ast::{SchemaDocumentExtension, TypeExtension};
use petgraph::{
    dot::Dot,
    graph::{EdgeIndex, NodeIndex},
    Directed, Graph as Petgraph,
};
use thiserror::Error;

use crate::{
    edge::Edge,
    node::Node,
    supergraph::{RootType, SupergraphDefinition, SupergraphMetadata},
};

type Graph = Petgraph<Node, Edge, Directed>;

#[derive(Debug)]
pub struct GraphQLSatisfiabilityGraph {
    pub lookup: LookupTable,
}

#[derive(Debug, Error)]
pub enum GraphQLSatisfiabilityGraphError {
    #[error("failed to parse schema: {0}")]
    ParseSchemaError(#[from] ParseError),
    #[error("failed to locate 'from' node with id {0}")]
    FromEdgeIdNotFound(String),
    #[error("failed to locate 'to' node with id {0}")]
    ToEdgeIdNotFound(String),
}

#[derive(Debug, Default)]
pub struct LookupTable {
    pub graph: Graph,
    pub query_root: NodeIndex,
    pub mutation_root: Option<NodeIndex>,
    pub subscription_root: Option<NodeIndex>,
    node_to_index: HashMap<String, NodeIndex>,
}

impl LookupTable {
    pub fn create_node_for_type(&mut self, node: Node) -> NodeIndex {
        let id = node.id();

        if self.node_to_index.contains_key(&id) {
            return *self.node_to_index.get(&id).unwrap();
        }

        let index = self.graph.add_node(node);
        self.node_to_index.insert(id, index);
        index
    }

    pub fn link_nodes_using_indices(
        &mut self,
        from_id: &str,
        to_id: &str,
        edge: Edge,
    ) -> Result<EdgeIndex, GraphQLSatisfiabilityGraphError> {
        let from = self.node_to_index.get(from_id).ok_or(
            GraphQLSatisfiabilityGraphError::FromEdgeIdNotFound(from_id.to_string()),
        )?;

        let to = self.node_to_index.get(to_id).ok_or(
            GraphQLSatisfiabilityGraphError::ToEdgeIdNotFound(from_id.to_string()),
        )?;

        Ok(self.link_nodes(*from, *to, edge))
    }

    pub fn link_nodes(&mut self, from: NodeIndex, to: NodeIndex, new_edge: Edge) -> EdgeIndex {
        let existing_edge = self.graph.edge_indices().find(|edge_index| {
            let (source, target) = self.graph.edge_endpoints(*edge_index).unwrap();
            let edge_weight = self.graph.edge_weight(*edge_index).unwrap();
            let is_same_nodes = source == from && target == to;
            let is_same_edge = edge_weight == &new_edge;

            is_same_nodes && is_same_edge
        });

        if let Some(edge) = existing_edge {
            edge
        } else {
            self.graph.add_edge(from, to, new_edge)
        }
    }
}

impl GraphQLSatisfiabilityGraph {
    pub fn new_from_supergraph(
        supergraph_ir: &SupergraphMetadata,
    ) -> Result<Self, GraphQLSatisfiabilityGraphError> {
        let mut lookup = LookupTable {
            node_to_index: HashMap::new(),
            graph: Graph::new(),
            ..Default::default()
        };

        lookup.query_root = lookup.create_node_for_type(Node::QueryRoot(
            supergraph_ir.document.query_type().name.clone(),
        ));
        lookup.mutation_root = supergraph_ir.document.mutation_type().map(|mutation_type| {
            lookup.create_node_for_type(Node::MutationRoot(mutation_type.name.clone()))
        });
        lookup.subscription_root =
            supergraph_ir
                .document
                .subscription_type()
                .map(|subscription_type| {
                    lookup.create_node_for_type(Node::SubscriptionRoot(
                        subscription_type.name.clone(),
                    ))
                });

        let mut instance = GraphQLSatisfiabilityGraph { lookup };

        instance.build_graph(supergraph_ir)?;

        Ok(instance)
    }

    fn build_graph(
        &mut self,
        supergraph_ir: &SupergraphMetadata,
    ) -> Result<(), GraphQLSatisfiabilityGraphError> {
        // First, we iterate all the definitions in the supergraph and build nodes for each of them.
        // We create all the nodes first, and then we create the edges.
        for definition in supergraph_ir.definitions.values() {
            Self::build_nodes_for_definition(&mut self.lookup, definition);
        }

        // Then, build the edges between the nodes.
        for definition in supergraph_ir.definitions.values() {
            // @join__type
            // First, we iterate the fields and find references across subgraphs
            // These needs to be linked with EntityReference edges, and basically represent the relation
            // between the same type in different subgraphs.
            Self::build_entity_reference_edges(&mut self.lookup, definition)?;
            // @join__field
            // Then, we iterate the fields and create edges between the fields and the field types.
            // These fields are simpler and considered local
            Self::build_field_edges(&mut self.lookup, definition)?;
            // @join__implements
            // Then, we iterate and build the edges that are based on interfaces.
            Self::build_interface_edges(&mut self.lookup, definition)?;
        }

        Ok(())
    }

    fn build_nodes_for_definition(lookup: &mut LookupTable, definition: &SupergraphDefinition) {
        for join_type in definition.join_types() {
            // We can skip creating Nodes for types that are not available in the current subgraph,
            // when iterating the root type.
            // We determine that by checking the fields of the object type and check what fields are involved in a join.
            if !definition.available_in_subgraph(&join_type.graph) && definition.is_root() {
                continue;
            }

            let node = Node::SubgraphType {
                name: definition.name().to_string(),
                subgraph: join_type.graph.to_string(),
            };

            // Create the Node for the type, and link it to the root if it's a root type.
            let node_index = lookup.create_node_for_type(node);

            match definition.root_type() {
                Some(RootType::Query) => {
                    lookup.link_nodes(lookup.query_root, node_index, Edge::Root);
                }
                Some(RootType::Mutation) => {
                    lookup.link_nodes(lookup.mutation_root.unwrap(), node_index, Edge::Root);
                }
                Some(RootType::Subscription) => {
                    lookup.link_nodes(lookup.subscription_root.unwrap(), node_index, Edge::Root);
                }
                _ => {}
            }

            // Iterate the fields of the object type and create nodes for them.
            // This will ensure smooth edges creation process, if all nodes already exists.
            for field in definition.fields().values() {
                let field_type = field.source.field_type.inner_type();

                // If a field does not have join field, it means that the field is local,
                // and we can create a node for it.
                if field.join_field.is_empty() {
                    let node = Node::SubgraphType {
                        name: field_type.to_string(),
                        subgraph: join_type.graph.to_string(),
                    };

                    lookup.create_node_for_type(node);
                } else {
                    // If a field has join field, we need to create a node for each subgraph it's available in.
                    for jf in &field.join_field {
                        if let Some(subgraph) = &jf.graph {
                            let node = Node::SubgraphType {
                                name: field_type.to_string(),
                                subgraph: subgraph.to_string(),
                            };

                            lookup.create_node_for_type(node);
                        }
                    }
                }
            }
        }
    }

    fn build_entity_reference_edges(
        lookup: &mut LookupTable,
        definition: &SupergraphDefinition,
    ) -> Result<(), GraphQLSatisfiabilityGraphError> {
        for join_type1 in definition.join_types() {
            for join_type2 in definition.join_types() {
                if join_type1.graph != join_type2.graph {
                    let name = definition.name();
                    let id_from = Node::id_from(name, Some(&join_type1.graph));
                    let id_to = Node::id_from(name, Some(&join_type2.graph));

                    if let Some(key) = &join_type2.key {
                        lookup.link_nodes_using_indices(
                            &id_from,
                            &id_to,
                            Edge::EntityReference(key.clone()),
                        )?;
                    }
                }
            }
        }

        Ok(())
    }

    fn build_interface_edges(
        lookup: &mut LookupTable,
        definition: &SupergraphDefinition,
    ) -> Result<(), GraphQLSatisfiabilityGraphError> {
        for join_implements in definition.join_implements() {
            let id_from = Node::id_from(definition.name(), Some(&join_implements.graph));
            let id_to = Node::id_from(&join_implements.interface, Some(&join_implements.graph));

            lookup.link_nodes_using_indices(
                &id_from,
                &id_to,
                Edge::InterfaceImplementation(join_implements.interface.clone()),
            )?;
        }

        Ok(())
    }

    fn build_field_edges(
        lookup: &mut LookupTable,
        definition: &SupergraphDefinition,
    ) -> Result<(), GraphQLSatisfiabilityGraphError> {
        for join_type in definition.join_types() {
            for (name, field) in definition.fields().iter() {
                let id_from = Node::id_from(definition.name(), Some(&join_type.graph));
                let type_to = field.source.field_type.inner_type();

                // If the field has no join field, we can link it directly to the field type.
                if field.join_field.is_empty() {
                    let id_to = Node::id_from(type_to, Some(&join_type.graph));

                    lookup.link_nodes_using_indices(
                        &id_from,
                        &id_to,
                        Edge::Field {
                            name: name.to_string(),
                            join_field: None,
                        },
                    )?;
                } else if definition.available_in_subgraph(&join_type.graph) {
                    for jf in &field.join_field {
                        if let Some(join_field_subgraph) = &jf.graph {
                            let subgraph_target = if *join_field_subgraph != join_type.graph {
                                join_field_subgraph
                            } else {
                                &join_type.graph
                            };

                            let id_to = Node::id_from(type_to, Some(subgraph_target));
                            lookup.link_nodes_using_indices(
                                &id_from,
                                &id_to,
                                Edge::Field {
                                    name: name.to_string(),
                                    join_field: Some(jf.clone()),
                                },
                            )?;
                        }
                    }
                }
            }
        }

        Ok(())
    }
}

/// Print me with `println!("{}", graph);` to see the graph in DOT/digraph format.
impl Display for GraphQLSatisfiabilityGraph {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:?}", Dot::with_config(&self.lookup.graph, &[]))
    }
}
