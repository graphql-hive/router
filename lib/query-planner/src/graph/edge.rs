use std::fmt::Debug;

use petgraph::graph::EdgeIndex;

use crate::federation_spec::directives::JoinFieldDirective;

use super::selection::SelectionNode;

pub type EdgePair<'a> = (&'a Edge, EdgeIndex);

pub enum Edge {
    /// A special edge between the root Node and then root entry point to the graph
    /// With this helper, you can jump from Query::RootQuery --field-> Query/SomeSubgraph --> --field--> SomeType/SomeSubgraph
    RootEntrypoint {
        field_name: String,
    },
    /// Represent a simple file move
    FieldMove {
        name: String,
        join_field: Option<JoinFieldDirective>,
        requires: Option<SelectionNode>,
        override_from: Option<String>,
    },
    EntityMove(String),
    /// join__implements
    AbstractMove(String),
    // interfaceObject
    // InterfaceObjectMove(String),
}

impl Edge {
    /// Helper to create a Field edge from a field name and join directive
    pub fn create_field_move(
        name: String,
        join_field: Option<JoinFieldDirective>,
        field_type: &str,
    ) -> Self {
        let requires = join_field.as_ref().and_then(|jf| jf.requires.clone());
        let override_from = join_field.as_ref().and_then(|jf| jf.override_value.clone());

        Self::FieldMove {
            name: name.clone(),
            join_field,
            requires: requires.map(|s| SelectionNode::parse_field_selection(s, &name, field_type)),
            override_from,
        }
    }

    pub fn id(&self) -> &str {
        match self {
            Self::FieldMove { name, .. } => name,
            Self::EntityMove(id) => id,
            Self::AbstractMove(id) => id,
            Self::RootEntrypoint { field_name } => field_name,
        }
    }

    /// Gets the requirements as a string, if any
    pub fn requirements(&self) -> Option<&SelectionNode> {
        match self {
            Self::FieldMove { requires, .. } => requires.as_ref(),
            _ => None,
        }
    }

    pub fn cost(&self) -> u64 {
        match self {
            Self::FieldMove { .. } => 1,
            _ => 10,
        }
    }
}

impl Debug for Edge {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Edge::RootEntrypoint { field_name } => write!(f, "root({})", field_name),

            Edge::FieldMove {
                name, join_field, ..
            } => {
                // Start with the field name
                let mut result = write!(f, "{}", name);

                // Add requires directive if present
                if let Some(jf) = join_field {
                    if let Some(req) = &jf.requires {
                        result = result.and_then(|_| write!(f, " @requires({})", req));
                    }

                    // Add provides directive if present
                    if jf.provides.is_some() {
                        result = result.and_then(|_| write!(f, " @provides"));
                    }

                    // Add other relevant directives like external, override, etc.
                    if jf.external.unwrap_or(false) {
                        result = result.and_then(|_| write!(f, " @external"));
                    }

                    if let Some(override_from) = &jf.override_value {
                        result =
                            result.and_then(|_| write!(f, " @override(from: {})", override_from));
                    }
                }

                result
            }
            Edge::EntityMove(name) => write!(f, "ðŸ”‘ {}", name),
            Edge::AbstractMove(name) => write!(f, "ðŸ”® {}", name),
        }
    }
}

impl PartialEq for Edge {
    fn eq(&self, other: &Self) -> bool {
        match (self, other) {
            (
                Edge::RootEntrypoint { field_name },
                Edge::RootEntrypoint {
                    field_name: other_field_name,
                },
            ) => field_name == other_field_name,
            (
                Edge::FieldMove {
                    name,
                    join_field: Some(jf1),
                    ..
                },
                Edge::FieldMove {
                    name: other_name,
                    join_field: Some(jf2),
                    ..
                },
            ) => {
                // Compare names and directive fields that affect planning
                name == other_name
                    && jf1.requires == jf2.requires
                    && jf1.provides == jf2.provides
                    && jf1.external == jf2.external
                    && jf1.override_value == jf2.override_value
            }

            (
                Edge::FieldMove {
                    name,
                    join_field: None,
                    ..
                },
                Edge::FieldMove {
                    name: other_name,
                    join_field: None,
                    ..
                },
            ) => name == other_name,

            (Edge::EntityMove(name), Edge::EntityMove(other_name)) => name == other_name,

            (Edge::AbstractMove(name), Edge::AbstractMove(other_name)) => name == other_name,

            _ => false,
        }
    }
}
