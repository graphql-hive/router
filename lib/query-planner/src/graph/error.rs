use crate::state::{
    selection_resolver::SelectionResolverError,
    supergraph_state::{OperationKind, SupergraphStateError},
};
use petgraph::graph::{EdgeIndex, NodeIndex};

#[derive(thiserror::Error, Debug)]
pub enum GraphError {
    #[error("Node with index '{0:?}' was not found")]
    NodeNotFound(NodeIndex),
    #[error("Edge with index '{0:?}' was not found")]
    EdgeNotFound(EdgeIndex),
    #[error("Unexpected missing root type {0}")]
    MissingRootType(OperationKind),
    #[error("Definition with name '{0}' was not found")]
    DefinitionNotFound(String),
    #[error("Field named '{0}' was not found in definition name '{1}'")]
    FieldDefinitionNotFound(String, String),
    #[error("Supergraph state error: {0}")]
    SupergraphStateError(Box<SupergraphStateError>),
    #[error("Selection resolver error: {0}")]
    SelectionResolverError(Box<SelectionResolverError>),
}

impl From<SupergraphStateError> for GraphError {
    fn from(err: SupergraphStateError) -> Self {
        GraphError::SupergraphStateError(Box::new(err))
    }
}

impl From<SelectionResolverError> for GraphError {
    fn from(err: SelectionResolverError) -> Self {
        GraphError::SelectionResolverError(Box::new(err))
    }
}
