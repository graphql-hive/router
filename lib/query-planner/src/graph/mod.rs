pub mod edge;
pub mod node;
mod tests;

use std::{
    collections::HashMap,
    fmt::{Debug, Display},
};

use graphql_parser_hive_fork::{
    query::{Selection, SelectionSet},
    schema::ParseError,
};
use graphql_tools::ast::{SchemaDocumentExtension, TypeExtension};
use node::SubgraphType;
use petgraph::{
    dot::Dot,
    graph::{EdgeIndex, Edges, NodeIndex},
    Directed, Direction, Graph as Petgraph,
};
use thiserror::Error;

use crate::{
    federation_spec::FederationRules,
    supergraph_metadata::{RootType, SupergraphDefinition, SupergraphState},
};

use super::graph::{edge::Edge, node::Node};

type Graph = Petgraph<Node, Edge, Directed>;

#[derive(Debug, Default)]
pub struct GraphQLSatisfiabilityGraph {
    pub graph: Graph,
    pub query_root: NodeIndex,
    pub mutation_root: Option<NodeIndex>,
    pub subscription_root: Option<NodeIndex>,
    pub node_to_index: HashMap<String, NodeIndex>,
}

#[derive(Debug, Error)]
pub enum GraphQLSatisfiabilityGraphError {
    #[error("failed to parse schema: {0}")]
    ParseSchemaError(#[from] ParseError),
    #[error("failed to locate 'from' node with id {0}")]
    FromEdgeIdNotFound(String),
    #[error("failed to locate 'to' node with id {0}")]
    ToEdgeIdNotFound(String),
}

impl GraphQLSatisfiabilityGraph {
    pub fn new_from_supergraph(
        supergraph_ir: &SupergraphState,
    ) -> Result<Self, GraphQLSatisfiabilityGraphError> {
        let mut instance = GraphQLSatisfiabilityGraph {
            node_to_index: HashMap::new(),
            graph: Graph::new(),
            ..Default::default()
        };

        instance.build_graph(supergraph_ir)?;

        Ok(instance)
    }

    pub fn node(&self, node_index: NodeIndex) -> &Node {
        self.graph.node_weight(node_index).unwrap()
    }

    pub fn edge(&self, edge_id: EdgeIndex) -> &Edge {
        self.graph.edge_weight(edge_id).unwrap()
    }

    fn build_graph(
        &mut self,
        state: &SupergraphState,
    ) -> Result<(), GraphQLSatisfiabilityGraphError> {
        self.build_root_nodes(state);
        self.link_root_edges(state);
        self.build_field_edges(state);
        self.build_interface_implementation_edges(state)?;
        self.build_entity_reference_edges(state)?;
        self.build_viewed_field_edges(state)?;

        Ok(())
    }

    fn build_root_nodes(&mut self, state: &SupergraphState<'_>) {
        self.query_root =
            self.upsert_node(Node::QueryRoot(state.document.query_type().name.clone()));
        self.mutation_root = state
            .document
            .mutation_type()
            .map(|mutation_type| self.upsert_node(Node::MutationRoot(mutation_type.name.clone())));
        self.subscription_root = state.document.subscription_type().map(|subscription_type| {
            self.upsert_node(Node::SubscriptionRoot(subscription_type.name.clone()))
        });
    }

    pub fn upsert_node(&mut self, node: Node) -> NodeIndex {
        let id = node.id();

        if self.node_to_index.contains_key(&id) {
            return *self.node_to_index.get(&id).unwrap();
        }

        let index = self.graph.add_node(node);
        self.node_to_index.insert(id, index);
        index
    }

    pub fn upsert_edge(&mut self, head: NodeIndex, tail: NodeIndex, edge: Edge) -> EdgeIndex {
        let existing_edge = self.graph.edge_indices().find(|edge_index| {
            let (source, target) = self.graph.edge_endpoints(*edge_index).unwrap();
            let edge_weight = self.graph.edge_weight(*edge_index).unwrap();
            let is_same_nodes = source == head && target == tail;
            let is_same_edge = edge_weight == &edge;

            is_same_nodes && is_same_edge
        });

        if let Some(edge) = existing_edge {
            edge
        } else {
            self.graph.add_edge(head, tail, edge)
        }
    }

    fn build_entity_reference_edges(
        &mut self,
        state: &SupergraphState<'_>,
    ) -> Result<(), GraphQLSatisfiabilityGraphError> {
        for (def_name, definition) in state.definitions.iter() {
            for join_type1 in definition.join_types() {
                for join_type2 in definition.join_types() {
                    let head =
                        self.upsert_node(Node::subgraph_type(def_name, &join_type1.graph_id));

                    if join_type1.graph_id != join_type2.graph_id {
                        if let Some(key) = &join_type2.key {
                            let tail = self
                                .upsert_node(Node::subgraph_type(def_name, &join_type2.graph_id));

                            self.upsert_edge(head, tail, Edge::EntityMove(key.clone()));
                        }
                    } else if let Some(key) = &join_type1.key {
                        self.upsert_edge(head, head, Edge::EntityMove(key.clone()));
                    }
                }
            }
        }

        Ok(())
    }

    fn build_interface_implementation_edges(
        &mut self,
        state: &SupergraphState<'_>,
    ) -> Result<(), GraphQLSatisfiabilityGraphError> {
        for (def_name, definition) in state
            .definitions
            .iter()
            .filter(|(_, d)| matches!(d, SupergraphDefinition::Object(_)))
        {
            for join_implements in definition.join_implements() {
                let tail =
                    self.upsert_node(Node::subgraph_type(def_name, &join_implements.graph_id));
                let head = self.upsert_node(Node::subgraph_type(
                    &join_implements.interface,
                    &join_implements.graph_id,
                ));

                self.upsert_edge(
                    head,
                    tail,
                    Edge::AbstractMove(definition.name().to_string()),
                );
            }
        }

        Ok(())
    }

    pub fn find_definition_node(
        &self,
        definition_name: &str,
        subgraph: &str,
    ) -> Option<(NodeIndex, &Node)> {
        let id = Node::id_from(definition_name, Some(subgraph));

        self.node_to_index
            .get(&id)
            .map(|&index| (index, &self.graph[index]))
    }

    pub fn root_query_node(&self) -> &Node {
        &self.graph[self.query_root]
    }

    pub fn root_mutation_node(&self) -> Option<&Node> {
        if let Some(mutation_root) = self.mutation_root {
            Some(&self.graph[mutation_root])
        } else {
            None
        }
    }

    pub fn root_subscription_node(&self) -> Option<&Node> {
        if let Some(subscription_root) = self.subscription_root {
            Some(&self.graph[subscription_root])
        } else {
            None
        }
    }

    pub fn edges_to(&self, node_index: NodeIndex) -> Edges<'_, Edge, Directed> {
        self.graph.edges_directed(node_index, Direction::Incoming)
    }

    pub fn edges_from(&self, node_index: NodeIndex) -> Edges<'_, Edge, Directed> {
        self.graph.edges_directed(node_index, Direction::Outgoing)
    }

    fn link_root_edges(&mut self, state: &SupergraphState<'_>) {
        for (def_name, definition) in state.definitions.iter() {
            if let Some(root_type) = definition.try_into_root_type() {
                for graph_id in definition.subgraphs().iter() {
                    for (field_name, field_definition) in definition.fields() {
                        let (is_available, _) = FederationRules::check_field_subgraph_availability(
                            field_definition,
                            graph_id,
                            definition,
                        );

                        if !is_available {
                            continue;
                        }

                        let head = match root_type {
                            RootType::Query => self.query_root,
                            RootType::Mutation => self.mutation_root.unwrap(),
                            RootType::Subscription => self.subscription_root.unwrap(),
                        };

                        let tail = self.upsert_node(Node::subgraph_type(def_name, graph_id));

                        self.upsert_edge(
                            head,
                            tail,
                            Edge::RootEntrypoint {
                                field_name: field_name.clone(),
                            },
                        );
                    }
                }
            }
        }
    }

    fn build_field_edges(&mut self, state: &SupergraphState<'_>) {
        for (def_name, definition) in state.definitions.iter() {
            for graph_id in definition.subgraphs().iter() {
                if definition.is_defined_in_subgraph(graph_id) {
                    for (field_name, field_definition) in definition.fields().iter() {
                        let (is_available, maybe_join_field) =
                            FederationRules::check_field_subgraph_availability(
                                field_definition,
                                graph_id,
                                definition,
                            );

                        let target_type = field_definition.source.field_type.inner_type();

                        match (is_available, maybe_join_field) {
                            (true, Some(join_field)) => {
                                let is_external = join_field.external.is_some_and(|v| v)
                                    && join_field.requires.is_none();
                                let has_provides = join_field.provides.is_some();

                                if is_external || has_provides {
                                    continue;
                                }

                                let head =
                                    self.upsert_node(Node::subgraph_type(def_name, graph_id));
                                let tail =
                                    self.upsert_node(Node::subgraph_type(target_type, graph_id));

                                self.upsert_edge(
                                    head,
                                    tail,
                                    Edge::create_field_move(
                                        field_name.clone(),
                                        Some(join_field.clone()),
                                    ),
                                );
                            }
                            (true, None) => {
                                let head =
                                    self.upsert_node(Node::subgraph_type(def_name, graph_id));
                                let tail =
                                    self.upsert_node(Node::subgraph_type(target_type, graph_id));
                                self.upsert_edge(
                                    head,
                                    tail,
                                    Edge::create_field_move(field_name.clone(), None),
                                );
                            }
                            // The field is not available in the current subgraph
                            _ => {}
                        };
                    }
                }
            }
        }
    }

    fn handle_viewed_selection_set(
        &mut self,
        state: &SupergraphState,
        selection_set: &SelectionSet<'static, String>,
        graph_id: &str,
        parent_type_def: &SupergraphDefinition<'_>,
        head: NodeIndex,
        view_id: u64,
    ) {
        for selection in selection_set.items.iter() {
            match selection {
                Selection::Field(field) => {
                    let is_leaf = field.selection_set.items.is_empty();
                    let field_in_parent = parent_type_def.fields().get(&field.name).unwrap();
                    let return_type_name = field_in_parent.source.field_type.inner_type();

                    let subgraph_type = SubgraphType {
                        name: return_type_name.to_string(),
                        subgraph: graph_id.to_string(),
                    };
                    let tail = self.upsert_node(match is_leaf {
                        true => Node::SubgraphType(subgraph_type),
                        false => Node::SubgraphTypeView {
                            view_id,
                            node: subgraph_type,
                            selection_set: field.selection_set.to_string(),
                        },
                    });

                    self.upsert_edge(
                        head,
                        tail,
                        Edge::create_field_move(field.name.to_string(), None),
                    );

                    if !is_leaf {
                        let return_type = state.definitions.get(return_type_name).unwrap();

                        self.handle_viewed_selection_set(
                            state,
                            &field.selection_set,
                            graph_id,
                            return_type,
                            tail,
                            view_id,
                        );
                    }
                }
                _ => unimplemented!("fragments are not supported in provides yet"),
            };
        }
    }

    fn build_viewed_field_edges(
        &mut self,
        state: &SupergraphState,
    ) -> Result<(), GraphQLSatisfiabilityGraphError> {
        for (_, definition) in state.definitions.iter() {
            for join_type in definition.join_types().iter() {
                let mut view_id = 0;

                for (field_name, field_definition) in definition.fields().iter() {
                    for join_field in field_definition.join_field.iter() {
                        if join_field
                            .graph_id
                            .as_ref()
                            .is_some_and(|v| v == &join_type.graph_id)
                            && join_field.provides.is_some()
                        {
                            if let Some(selection_set) = FederationRules::parse_provides(join_field)
                            {
                                view_id += 1;

                                let head = self.upsert_node(Node::subgraph_type(
                                    definition.name(),
                                    &join_type.graph_id,
                                ));

                                let return_type_name =
                                    field_definition.source.field_type.inner_type();

                                let tail = self.upsert_node(Node::SubgraphTypeView {
                                    view_id,
                                    node: SubgraphType {
                                        name: return_type_name.to_string(),
                                        subgraph: join_type.graph_id.to_string(),
                                    },
                                    selection_set: selection_set.to_string(),
                                });

                                self.upsert_edge(
                                    head,
                                    tail,
                                    Edge::create_field_move(
                                        field_name.to_string(),
                                        Some(join_field.clone()),
                                    ),
                                );

                                let return_type = state.definitions.get(return_type_name).unwrap();

                                self.handle_viewed_selection_set(
                                    state,
                                    &selection_set,
                                    &join_type.graph_id,
                                    return_type,
                                    tail,
                                    view_id,
                                );
                            }
                        }
                    }
                }
            }
        }

        Ok(())
    }
}

/// Print me with `println!("{}", graph);` to see the graph in DOT/digraph format.
impl Display for GraphQLSatisfiabilityGraph {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:?}", Dot::with_config(&self.graph, &[]))
    }
}
