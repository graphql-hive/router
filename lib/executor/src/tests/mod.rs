use query_planner::graph::PlannerOverrideContext;
use sonic_rs::JsonValueTrait;

use crate::response::graphql_error::GraphQLErrorPathSegment;
use crate::{
    context::QueryPlanExecutionContext, execution::plan::QueryPlanExecutor,
    executors::common::SubgraphExecutor, introspection::schema::SchemaWithMetadata,
    SubgraphExecutorMap,
};

mod async_graphql;
mod fixtures;

#[test]
fn error_propagation() {
    let supergraph_sdl =
        std::fs::read_to_string("./src/tests/fixtures/error_propagation/supergraph.graphql")
            .expect("Unable to read input file");
    let parsed_schema = query_planner::utils::parsing::parse_schema(&supergraph_sdl);
    let planner = query_planner::planner::Planner::new_from_supergraph(&parsed_schema)
        .expect("Failed to create planner from supergraph");
    let parsed_document = query_planner::utils::parsing::parse_operation(
        &std::fs::read_to_string("./src/tests/fixtures/error_propagation/operation.graphql")
            .expect("Unable to read input file"),
    );
    let normalized_document = query_planner::ast::normalization::normalize_operation(
        &planner.supergraph,
        &parsed_document,
        None,
    )
    .expect("Failed to normalize operation");
    let normalized_operation = normalized_document.executable_operation();
    let query_plan = planner
        .plan_from_normalized_operation(normalized_operation, PlannerOverrideContext::default())
        .expect("Failed to create query plan");

    let schema_metadata = SchemaWithMetadata::schema_metadata(&planner.consumer_schema);
    let movies_subgraph = fixtures::error_propagation::movies::get_subgraph();
    let directors_subgraph = fixtures::error_propagation::directors::get_subgraph();
    let mut subgraph_executor_map = SubgraphExecutorMap::new();
    subgraph_executor_map.insert_boxed_arc(
        "movies".to_string(),
        SubgraphExecutor::to_boxed_arc(movies_subgraph),
    );
    subgraph_executor_map.insert_boxed_arc(
        "directors".to_string(),
        SubgraphExecutor::to_boxed_arc(directors_subgraph),
    );
    tokio_test::block_on(async {
        let qp_executor = QueryPlanExecutor::new(&None, &subgraph_executor_map, &schema_metadata);
        let mut qp_exec_ctx =
            QueryPlanExecutionContext::new(&query_plan, crate::response::value::Value::Null);
        qp_executor
            .execute(&mut qp_exec_ctx, query_plan.node.as_ref())
            .await;
        assert_eq!(qp_exec_ctx.errors.len(), 1);
        let error = &qp_exec_ctx.errors[0];
        assert_eq!(
            error.path,
            Some(vec![GraphQLErrorPathSegment::String("movie2".to_string())])
        );
        assert_eq!(error.message, "Director not found for movie with id 2");
        assert_eq!(
            error
                .extensions
                .as_ref()
                .map(|ext| ext.get("code").map(|v| v.as_str()))
                .flatten()
                .flatten(),
            Some("DOWNSTREAM_SERVICE_ERROR")
        );
        assert_eq!(
            error
                .extensions
                .as_ref()
                .map(|ext| ext.get("serviceName").map(|v| v.as_str()))
                .flatten()
                .flatten(),
            Some("directors")
        );
        insta::assert_snapshot!(qp_exec_ctx.final_response.to_string());
    });
}
