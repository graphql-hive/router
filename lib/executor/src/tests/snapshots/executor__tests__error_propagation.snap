---
source: lib/executor/src/tests/mod.rs
expression: qp_exec_ctx.final_response.to_string()
---
{"movie1": {"__typename": "Movie", "director": {"id": "1", "name": "Christopher Nolan"}, "id": "1", "name": "Inception"}, "movie2": {"__typename": "Movie", "id": "2", "name": "Interstellar"}}
